pennylane/optimize/rotoselect.py

```diff
@@ -0,0 +1,189 @@
+# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Rotoselect gradient free optimizer"""
+
+import pennylane as qml
+from pennylane import numpy as np
+from pennylane.utils import _flatten, unflatten
+
+
+class RotoselectOptimizer:
+    r"""Rotoselect gradient free optimizer.
+
+    The Rotoselect optimizer minimizes an objective function with respect to the parameters and
+    gates of a quantum circuit without the need of calculating the gradient of the function. The
+    algorithm works by updating the parameters :math:`\theta = \theta_1, \dots, \theta_D` and gate
+    choices :math:`R = R_1,\dots,R_D` one at a time according to a closed-form expression for the
+    optimal value of the :math:`d^{th}` parameter :math:`\theta^*_d` when the other parameters and
+    gate choices are fixed:
+
+    .. math:: \theta^*_d = \underset{\theta_d}{\text{argmin}}\left<H\right>_{\theta_d}
+              = -\frac{\pi}{2} - \text{arctan2}\left(2\left<H\right>_{\theta_d=0}
+              - \left<H\right>_{\theta_d=\pi/2} - \left<H\right>_{\theta_d=-\pi/2},
+              \left<H\right>_{\theta_d=\pi/2} - \left<H\right>_{\theta_d=-\pi/2}\right),
+
+    where :math:`\left<H\right>_{\theta_d}` is the expectation value of the objective function
+    optimized over the parameter :math:`\theta_d`. :math:`\text{arctan2}(x, y)` computes the
+    element-wise arc tangent of :math:`x/y` choosing the quadrant correctly, avoiding, in
+    particular, division-by-zero when :math:`y = 0`.
+
+    Which parameters and gates that should be optimized over is decided in the user-defined cost
+    function, where :math:`R` would be passed on as gates in a quantum circuit, along
+    with the parameters :math:`\theta` for the circuit and its gates. Note that not all gates
+    or paramaters need to be optimized over.
+
+    The algorithm is described in further detail in `Ostaszewski et al. (2019) <https://arxiv.org/abs/1905.09692>`_.
+
+    Keyword Args:
+        possible_generators (list[~.Operation]): List containing the possible ``pennylane.ops.qubit`` operators
+            that are allowed in the circuit. Default is the set of Pauli matrices :math:`\{X, Y, Z\}`.
```

Comment:
Should this be Pauli matrix rotations?
```suggestion
            that are allowed in the circuit. Default is the set of Pauli matrices :math:`\{R_x, R_y, R_z\}`.
```

Reply:
I guess the confusion here is related to Maria's comment https://github.com/XanaduAI/pennylane/pull/539/files#r390771617

Reply:
Ah, yes, thanks for pointing it out.